### PR TITLE
Feature/#316 cosine

### PR DESCRIPTION
**Pull request by davidpig**
The COSINE trainer is added.

**How to use it?**
- `knodle\examples\trainer\cosine\cosine_training_example.py` contains an example code for training and testing cosine on the TREC dataset. The TREC dataset is downloaded and preprocessed in Knodle dataset format.
- All cosine related code components can be found in `knodle\knodle\trainer\cosine`
- After the training is completed  (i.e. after `trainer.train()`), the best model during training obtained via early stopping is stored in `trainer.model`.

**Remarks**
- training with COSINE on the TREC dataset (via weak supervision) is tested.
- The function `convert_text_to_transformer_input()` in `knodle\examples\trainer\preprocessing.py` is updated, now it takes an additional parameter `max_sen_len`

### CODE REVIEW

File: knodle/trainer/cosine/cosine.py [hunk 1]
```diff
@@ -0,0 +1,496 @@
...
+        # split to labeled and unlabeled
+        l_set, ul_set = train_set.get_covered_subset()
+
+        # create validation set
+        dev_x_dict = {'input_ids': self.dev_model_input_x.tensors[0], 'attention_mask': self.dev_model_input_x.tensors[1]}
+        val_set = CosineDataset(dev_x_dict, self.dev_gold_labels_y.tensors[0])
+
+        return l_set, ul_set, val_set
+
+    def concat_datasets(self, set1, set2):
```
Comment (AndSt) [inline knodle/trainer/cosine/cosine.py:125]: could be marked as static method

File: knodle/trainer/cosine/cosine.py [hunk 2]
```diff
@@ -0,0 +1,496 @@
...
+            weight = torch.sum(-torch.log(target + 1e-6) * target, dim=1)  # Entropy
+            weight = 1 - weight / np.log(weight.size(-1))
+            # w = torch.FloatTensor([1 if x == True else 0 for x in weight > thresh]).to(device)
+            # w = torch.tensor(weight > thresh, dtype=torch.float)
+            w = (weight > thresh).float()
+        else:
+            raise ValueError(f'conf={conf} is unsupported')
+        target = self.soft_frequency(target, probs=True, soft=soft)
+
+        loss_batch = loss(input, target)
```
Comment (AndSt) [inline knodle/trainer/cosine/cosine.py:216]: You compute it on the full batch. Shouldn't this be computed on the subset "C" (samples over treshold, formula 10 in paper)?

File: knodle/trainer/cosine/cosine.py [hunk 3]
```diff
@@ -0,0 +1,496 @@
...
+
+        lc = torch.sum(loss_batch * w.unsqueeze(1) * weight.unsqueeze(1))  # l_c loss in the paper
+
+        n_classes_ = input.shape[-1]
+        # Note this is l-=, i.e l = l - (...)
+        lc -= confreg * (torch.sum(input * w.unsqueeze(1)) + np.log(n_classes_) * n_classes_)  # l_c + \lambda * R2 loss in the paper
```
Comment (AndSt) [inline knodle/trainer/cosine/cosine.py:222]: So this to me is formula 12: I think in the end it should be 1 / n_classes_ and I'm not sure whether they use the weights w here. After calculation I arrive here:
`
lc -= confreg * (torch.mean(  torch.sum(torch.log(input) + torch.log(n_classes_))  / n_classes_)
`
with input \in (batch, num_classes), n_classes \in 1

File: knodle/trainer/cosine/EarlyStopper.py [hunk 4]
```diff
@@ -0,0 +1,93 @@
...
+                            Default: print
+        """
+        self.patience = patience
+        self.verbose = verbose
+        self.counter = 0
+        self.best_score = None
+        self.early_stop = False
+        self.delta = delta
+        self.save_dir = save_dir
+        self.trace_func = trace_func
```
[the remark this answers is not in the document]
Comment (davidpig) [reply knodle/trainer/cosine/EarlyStopper.py:33]: Hi,
what do you mean by "logging" here?
no, I don't have anything special in my mind. This is just my own implementation of the early stopping used in my own projects.

File: knodle/trainer/cosine/cosine_utils.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        else:
+            print("Error Loss Metric!!")
+            return 0
+        #dist = torch.sum( - x0 * x1 / np.sqrt(x0.shape[-1]), 1).exp()
+        #dist_sq = dist ** 2
+
+        mdist = self.margin - dist
+        dist = torch.clamp(mdist, min=0.0)
+        loss = y * dist_sq + (1 - y) * torch.pow(dist, 2)
+        loss = torch.sum(loss) / 2.0 / x0.size()[0]
```
[the remark this answers is not in the document]
Comment (davidpig) [reply knodle/trainer/cosine/cosine_utils.py:49]: Hi, thanks for checking the details in the code.
ok, first of all. This part is from the original implementation, I hadn't checked it in detail.
1. According to footnote 4 in the paper, they applied "scaled Euclidean distance", hence `/2.0`
2. I guess they use `x0.size()[0]` to get the average loss of a batch, maybe `loss = torch.mean(loss)/2.0` will do it too.

File: knodle/trainer/cosine/cosine.py [hunk 6]
```diff
@@ -0,0 +1,496 @@
...
+from knodle.trainer.cosine.config import CosineConfig
+from knodle.trainer.utils.denoise import activate_neighbors
+
+logger = logging.getLogger(__name__)
+
+
+# This code (cosine trainer class) is adapted from the
+# official code of the COSINE framework (https://github.com/yueyu1030/COSINE)
+
+class MajorityVoting:
```
[the remark this answers is not in the document]
Comment (davidpig) [reply knodle/trainer/cosine/cosine.py:42]: sorry, I cannot really remember it.
I guess it's something related to how to resolve the weak rules if two or more weak rules contradict each other.

File: knodle/trainer/cosine/cosine.py (hunk 2 above)
Comment (davidpig) [reply knodle/trainer/cosine/cosine.py:216]: Hi,
yes, the variable `loss_batch` is indeed computed over the entire batch. However, the final loss `lc` is computed on the subset "C" (line 218 in the current code version).

Comment (AndSt) [reply knodle/trainer/cosine/cosine.py:216]: Hi, thanks for clarification.  I think I had problems understanding w / weight. This and line 218 makes sense to me now.
